### PR TITLE
Add support for blackboxes in FAMETransform via clock gating

### PR DESCRIPTION
**Pull request by albert-magyar**

### CODE REVIEW

File: src/main/scala/midas/Compiler.scala [hunk 1]
```diff
@@ -58,6 +58,7 @@ object MidasCompiler {
      (implicit p: Parameters): CircuitState = {
     val midasAnnos = Seq(
       firrtl.TargetDirAnnotation(dir.getPath()),
+      firrtl.transforms.BlackBoxTargetDirAnno(dir.getPath()),
```
Comment (davidbiancolin) [inline src/main/scala/midas/Compiler.scala]: This needs to go in the GoldenGateCompilerPhase. The MidasCompiler companion object shouldn't be used anymore. I'm assuming you haven't pulled in the stage stuff?

File: src/main/scala/midas/passes/fame/FAMETransform.scala [hunk 2]
```diff
@@ -177,17 +187,23 @@ object FAMEModuleTransformer {
         inChannelMap.getOrElse(name, outChannelMap(name)).replacePortRef(iWR)
       case oWR @ WRef(name, tpe, PortKind, FEMALE) if tpe != ClockType =>
         outChannelMap(name).replacePortRef(oWR)
+      case wr: WRef if wr.name == hostClock.name =>
+        // Replace host clock references with target clock references
+        targetClock.ref
       case e => e
     }
 
+    // A target state trigger is only needed if clocks are not gated
```
Comment (davidbiancolin) [inline src/main/scala/midas/passes/fame/FAMETransform.scala]: Maybe add a comment here explaining that CP removes the added clock-gating circuitry under this case?

File: src/main/scala/midas/passes/fame/FAMEUtils.scala [hunk 3]
```diff
@@ -56,7 +59,14 @@ private[fame] class FAMEChannelAnalysis(val state: CircuitState, val fameType: F
   // TODO: only transform submodules of model modules
   // TODO: add renames!
   val circuit = state.circuit
-  val syncModules = circuit.modules.filter(_.ports.exists(_.tpe == ClockType)).map(m => ModuleTarget(circuit.main, m.name)).toSet
+  val topTarget = ModuleTarget(circuit.main, circuit.main)
+  def moduleTarget(m: DefModule) = topTarget.copy(module = m.name)
+  def moduleTarget(wi: WDefInstance) = topTarget.copy(module = wi.module)
+
+  def stateful(m: DefModule) = m.ports.exists(_.tpe == ClockType)
+  // Don't include blackboxes
```
Comment (davidbiancolin) [inline src/main/scala/midas/passes/fame/FAMEUtils.scala]: Remove this comment or add "have no clock ports/are stateless"

File: src/main/scala/midas/passes/package.scala [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+import firrtl.WrappedExpression._
+
+import annotations._
+import TargetToken.{Instance, OfModule}
+import firrtl.Utils.BoolType
+import firrtl.transforms.BlackBoxInlineAnno
+
+import scala.language.implicitConversions
+
+
```
Comment (davidbiancolin) [inline src/main/scala/midas/passes/package.scala:16]: There's a bunch of great sugar in this Package object, it would be nice to see a little bit of Scala doc on it.

File: src/main/scala/midas/Compiler.scala (hunk 1 above)
Comment (davidbiancolin) [reply src/main/scala/midas/Compiler.scala]: https://github.com/ucb-bar/midas/blob/dev/src/main/scala/midas/stage/GoldenGateCompilerPhase.scala#L35

**Review by davidbiancolin — CHANGES_REQUESTED** [review]
Just need to move the annotations into the Phase. Otherwise this looks good!

File: src/main/scala/midas/Compiler.scala (hunk 1 above)
Comment (albert-magyar) [reply src/main/scala/midas/Compiler.scala]: I added that annotation before I merged in the Golden Gate stage, but I have been using the stage (this tree is based on that), and didn't have any problems when I merged that in last week. So I was a little confused when I saw what you pointed out.

As it turns out, using a stage at all actually changes the default blackbox resource target directory from its [previous default of the working directory](https://github.com/freechipsproject/firrtl/blob/973ecf516c0ef2b222f2eb68dc8b514767db59af/src/main/scala/firrtl/transforms/BlackBoxSourceHelper.scala#L59) to [a new default of the the output target directory](https://github.com/freechipsproject/firrtl/blob/a7cf6ff3416a11088d811a435ba71fd36b191fb4/src/main/scala/firrtl/stage/phases/AddDefaults.scala#L30) from the working directory to , so merging in the stage simultaneously prevented this annotation from being used while also obviating the need for it entirely, so I didn't really notice anything.

TLDR: I just deleted this line entirely and all is well.

File: src/main/scala/midas/passes/fame/FAMEUtils.scala (hunk 3 above)
Comment (albert-magyar) [reply src/main/scala/midas/passes/fame/FAMEUtils.scala]: The line of code below the comment filters out the blackboxes, so I left in the "don't include blackboxes," but I expanded the comments.

File: src/main/scala/midas/passes/fame/FAMETransform.scala (hunk 2 above)
Comment (albert-magyar) [reply src/main/scala/midas/passes/fame/FAMETransform.scala]: K

File: src/main/scala/midas/passes/package.scala (hunk 4 above)
Comment (albert-magyar) [reply src/main/scala/midas/passes/package.scala:16]: Added

File: src/main/scala/midas/passes/package.scala [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+    }
+
+    /**
+      * Examine the circuit and store information in an analysis object
+      */
+    def analyze(cs: CircuitState): T
+
+    /**
+      * @param analysis The results of the analysis pass
+      * @return Returns a function that, when called, transformes the circuit
```
Comment (davidbiancolin) [inline src/main/scala/midas/passes/package.scala]:
```suggestion
      * @return Returns a function that, when called, transforms the circuit
```

File: src/main/scala/midas/passes/package.scala [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+      */
+    def transformer(analysis: T): Circuit => Circuit
+
+    /**
+      * @param analysis The results of the analysis pass
+      * @return Returns a function that, when called, transformes the annotations
```
Comment (davidbiancolin) [inline src/main/scala/midas/passes/package.scala:220]:
```suggestion
      * @return Returns a function that, when called, transforms the annotations
```

**Review by davidbiancolin — APPROVED** [review]
LGTM

File: src/main/scala/midas/passes/fame/FAMETransform.scala [hunk 7]
```diff
@@ -177,17 +187,31 @@ object FAMEModuleTransformer {
...
     }
 
+   /*
+    * A target state trigger is only needed if clocks are not gated.  When using true clock gating,
+    * the transform still programmatically adds an extra enable signal to the state updates, so we
+    * pass in a constant one as the value of this enable signal.  This spurious condition gets
+    * optimized away during ConstantPropagation. This homogeneity is helpful since a transformed
+    * module might be instantiated within multiple top-level simulation models, some of which may
+    * rely true clock gating (if their corresponding target modules contain blackboxes) and some of
+    * which may not.
```
Comment (davidbiancolin) [inline src/main/scala/midas/passes/fame/FAMETransform.scala:203]: Cool, thanks for this expanded explanation.
